src/bank02.asm

```diff
@@ -8130,29 +8126,197 @@ titleScreenLicenseText:
-    TXT  "  every thing of<00>"                        ;; 02:7ec3 .................
-    TXT  "    this world.<00>"                         ;; 02:7ed4 ................
-    TXT  "<00>"                                        ;; 02:7ee4 ?
-    TXT  "  It grows high<00>"                         ;; 02:7ee5 ????????????????
-    TXT  " above the clouds<00>"                       ;; 02:7ef5 ??????????????????
-    TXT  "in the air on top<00>"                       ;; 02:7f07 ??????????????????
-    TXT  "of Mount Illusia.<00>"                       ;; 02:7f19 ??????????????????
-    TXT  "<00>"                                        ;; 02:7f2b ?
-    TXT  "Legend tells that<00>"                       ;; 02:7f2c ??????????????????
-    TXT  "it gives eternal<00>"                        ;; 02:7f3e ?????????????????
-    TXT  "power to the one<00>"                        ;; 02:7f4f ?????????????????
-    TXT  " who touched it.<00>"                        ;; 02:7f60 ?????????????????
-    TXT  "<00>"                                        ;; 02:7f71 ?
-    TXT  "  Dark Lord was<00>"                         ;; 02:7f72 ????????????????
-    TXT  "  trying to find<00>"                        ;; 02:7f82 ????????????????.
-    TXT  "  the way to the<00>"                        ;; 02:7f93 ...........??????
-    TXT  " Tree of Mana to<00>"                        ;; 02:7fa4 ?????????????????
-    TXT  "  get the mighty<00>"                        ;; 02:7fb5 ?????????????????
-    TXT  " power to conquer<00>"                       ;; 02:7fc6 ??????????????????
-    TXT  "    the world.<00>"                          ;; 02:7fd8 ???????????????
-    TXT  "<00>"                                        ;; 02:7fe7 ?
-    db   $00, $00, $00, $00, $00, $00, $00, $00        ;; 02:7fe8 ????????
-    db   $00, $00, $00, $00, $01, $ff, $ff, $ff        ;; 02:7ff0 ????????
-    db   $ff, $ff, $ff, $ff, $ff, $ff, $ff, $ff        ;; 02:7ff8 ????????
+    TXT  "<00>"                                        ;; 02:7fd8 ?
+    db   $01                                           ;; 02:7fd9 ?
+
+; Draws left aligned number HL as modified tiles at WRAM position DE
+; C holds the drawing mode in the first nibble (0 for shift, 1 for shift/swap)
+; C holds number of max digits in the second nibble
+drawLeftAlignedNumberInWRAM:
+    ld   B, $00 ; digit counter
+
+    ; Housekeeping, get a copy of proper WRAM starting
+    ;  location on the stack for later use
+    push BC
+    push DE
+
+    ; Load up the digits on the stack
+.divmod_and_store_number:
+    ld   A, $0a
+    push BC
+    push DE
+    call divMod
+    pop  DE
+    pop  BC
+    push AF
+    inc  B
+    ld   A, H
+    or   A, L
+    jr   NZ, .divmod_and_store_number
+
+    ; Reduce second nibble of C by num digits found
+    ; Remainder provides the number of tiles to clear
+    ; after we write the digits
+    ld   A, C
+    sub  A, B
+    ld   C, A
+
+    ; Increment tile buffer pointer (DE) if in swap mode
```

Comment:
Okay, so this is because when swapping the final number of tiles is equal to the number of digits *plus one*, and the swap goes from left to right, so if a swap is to be performed then the first tile needs to be untouched by the shift?

Reply:
Yes, mostly. The first tile needs to be empty, it's cleared later in the function. So before the swap routine we have, in order:

- An empty tile
- (# of digits) tiles with left aligned digits
- 3-(# of digits) tiles cleared out

Then the swap (shift_4px) routine starts at the empty tile and works its way forward, moving the contents of the tiles 4 pixels to the left across the tile boundaries.

Reply:
I added a bit more commenting to make this clearer. I also moved the first tile clear to this area. Check it out and see if it reads easier.